extensions/panache/panache-mock/src/main/java/io/quarkus/panache/mock/impl/PanacheQuarkusComponetTestCallbacks.java

```diff
@@ -0,0 +1,189 @@
+                    if (userMethods == null) {
+                        userMethods = new ArrayList<>();
+                        entityUserMethods.put(entity.name(), userMethods);
+                    }
+                    userMethods.add(method);
+                }
+            }
+        }
+
+        Map<DotName, List<MethodInfo>> entityGeneratedMethods = new HashMap<>();
+        ClassInfo panacheEntityBaseClassInfo = buildContext.getComputingBeanArchiveIndex()
+                .getClassByName(DotName.createSimple("io.quarkus.hibernate.orm.panache.PanacheEntityBase"));
+        for (ClassInfo entity : mockedEntities) {
+            for (MethodInfo method : panacheEntityBaseClassInfo.methods()) {
+                if (!userMethodExists(entityUserMethods.get(entity.name()), method)) {
+                    AnnotationInstance bridge = method.annotation(PanacheConstants.DOTNAME_GENERATE_BRIDGE);
+                    if (bridge != null) {
+                        // TODO bridge.value("targetReturnTypeErased") and bridge.value("callSuperMethod")
+                        List<MethodInfo> generated = entityGeneratedMethods.get(entity.name());
+                        if (generated == null) {
+                            generated = new ArrayList<>();
+                            entityGeneratedMethods.put(entity.name(), generated);
+                        }
+                        generated.add(method);
+                    }
+                }
+            }
+        }
+
+        for (ClassInfo entity : mockedEntities) {
+            String entityClassName = entity.name().toString();
+            ClassTransformer transformer = new ClassTransformer(entity.name().toString());
+            List<MethodInfo> userMethods = entityUserMethods.get(entity.name());
+            if (userMethods != null) {
+                for (MethodInfo userMethod : userMethods) {
+                    transformer.removeMethod(MethodDescriptor.of(userMethod));
+                    addMethod(entityClassName, transformer, userMethod);
+                }
+            }
+            List<MethodInfo> generatedMethods = entityGeneratedMethods.get(entity.name());
+            if (generatedMethods != null) {
+                for (MethodInfo generatedMethod : generatedMethods) {
+                    addMethod(entityClassName, transformer, generatedMethod);
+                }
+            }
+            buildContext.addBytecodeTransformer(
+                    new BytecodeTransformer(entityClassName, new BiFunction<String, ClassVisitor, ClassVisitor>() {
+                        @Override
+                        public ClassVisitor apply(String className, ClassVisitor originalVisitor) {
+                            return transformer.applyTo(originalVisitor);
+                        }
+                    }));
+        }
+    }
+
+    private void addMethod(String entityClass, ClassTransformer transformer, MethodInfo method) {
```

Comment:
What's the rationale for this additional mechanism? Is it because for some reason `PanacheMockMethodCustomizer` cannot be invoked?

Reply:
I have to admit I don't know what `QuarkusComponentTest` does and how it differs from the other tests?

Reply:
Ah. I see. But, if transformers are not called, not even https://quarkus.io/guides/logging#simplified-logging ?

Reply:
That is correct -- simplified logging has a special case for test code (naively detected by presence of a JUnit 5 class) where it uses `StackWalker` to figure out the caller.

Reply:
Oh, wow.

Reply:
So wait, actually… we could make this work similarly, but having some code in the Panache methods instead of the `throws implementationMissing()`.

Then users would not even have to add `@MockPanacheEntities`.

Something like:

```java
    @GenerateBridge(targetReturnTypeErased = true)
    public static <T extends PanacheEntityBase> T findById(Object id) {
        return SomeClass.something(id);
    }
```

```java
class SomeClass {
 static Object something(Object... whatever) {
  if(componentTestingDetected()) {
    return mockPanacheMethodUsingStackWalker(whatever);
  }
  throw implementationInjectionMissing();
 }
}
```

No? Wouldn't this be even easier for users?

Reply:
Wait a minute. Right now, we copy the static methods from `PanacheEntityBase` to `SomeEntity` and then modify the bytecode of those methods to use `PanacheMock` instead. How and where would you like to use the `StackWalker`?

Reply:
You do this because you require `@MockPanacheEntities`, but if we add an API hook directly in `PanacheEntityBase` we can make it work in tests without any user setup.

Reply:
> You do this because you require `@MockPanacheEntities`, but if we add an API hook directly in `PanacheEntityBase` we can make it work in tests without any user setup.

I'm sorry but I don't understand how. 

Panache logging is IMO very different - the static methods exists and are called from the user code. So it's easy to find the caller with `StackWalker` and use the correct logger. 

In case of Panache entities, the user code would call (usually inherited static method from `PanacheEntityBase`) and the caller does not really help because it's not the entity class. How would you identify the correct mocked entity?

Reply:
OK, very good point. You're right :)
Ooops.